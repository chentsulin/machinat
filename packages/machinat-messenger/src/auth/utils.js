// @flow
/* eslint-disable import/prefer-default-export */
import type { AuthorizerRefineResult } from '@machinat/auth/types';
import type { ExtensionContext } from '../types';
import MessengerChannel from '../channel';
import MessengerUser from '../user';

export const refineExtensionContext = (
  ctx: ExtensionContext
): null | AuthorizerRefineResult => {
  if (!ctx || !ctx.page_id || !ctx.psid || !ctx.tid) {
    return null;
  }

  return {
    user: new MessengerUser(ctx.page_id, ctx.psid),
    sourceChannel: MessengerChannel.fromExtensionContext(ctx),
  };
};
