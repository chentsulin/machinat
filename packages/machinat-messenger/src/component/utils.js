import invariant from 'invariant';
import { MESSENGER_NAITVE_TYPE } from '../symbol';

export const annotateNative = component =>
  Object.assign(component, { $$native: MESSENGER_NAITVE_TYPE });

export const annotateNativeRoot = (component, entry) =>
  Object.assign(component, {
    $$entry: entry,
    $$native: MESSENGER_NAITVE_TYPE,
    $$root: true,
  });

const textReducer = (text, cur) => {
  invariant(
    typeof cur.rendered === 'string',
    `expect element at ${cur.path} can be rendered as string, got ${
      cur.element
    }`
  );
  return text + cur.rendered;
};

export const renderTextContent = (node, render, propName) => {
  const rendered = render(node, `.${propName}`);
  if (rendered === undefined) {
    return undefined;
  }

  return rendered.reduce(textReducer, '');
};

export const getRendered = result => result.rendered;
export const getElement = result => result.element;

const onlyType = (Type, result) => {
  for (let i = 0; i < result.length; i += 1) {
    if (Type !== result[i].element.type) return false;
  }
  return true;
};
export const renderOnlyType = (Type, node, render, propName) => {
  const result = render(node, `.${propName}`);
  if (result === undefined) {
    return undefined;
  }

  invariant(
    onlyType(Type, result),
    `${propName} prop should contain only ${
      Type.name
    } elements, got ${result.map(getElement)}`
  );
  return result;
};

const onlyInTypes = (types, result) => {
  for (let i = 0; i < result.length; i += 1) {
    if (types.indexOf(result[i].element.type) === -1) return false;
  }
  return true;
};
export const renderOnlyInTypes = (types, node, render, propName) => {
  const result = render(node, `.${propName}`);
  if (result === undefined) {
    return undefined;
  }

  invariant(
    onlyInTypes(types, result),
    `${propName} prop should contain only ${types
      .map(Type => Type.name)
      .join(', ')} elements, got ${result.map(getElement)}`
  );
  return result;
};
