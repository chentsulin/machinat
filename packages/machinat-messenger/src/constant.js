// @flow
export const MESSENGER = 'messenger';
export const MESSENGER_NAMESPACE = 'Messenger';
export const MESSENGER_NATIVE_TYPE = Symbol.for('machinat.messenger.native');

export const ATTACHED_FILE_DATA = Symbol(
  'machinat.messenger.job.attached_file_data'
);

export const ATTACHED_FILE_INFO = Symbol(
  'machinat.messenger.job.attached_file_info'
);

export const ENTRY_MESSAGES = 'me/messages';

export const ENTRY_PASS_THREAD_CONTROL = 'me/pass_thread_control';
export const ENTRY_REQUEST_THREAD_CONTROL = 'me/request_thread_control';
export const ENTRY_TAKE_THREAD_CONTROL = 'me/take_thread_control';

export const ENTRY_MESSAGE_CREATIVES = 'me/message_creatives';
export const ENTRY_BROADCAST_MESSAGES = 'me/broadcast_messages';
export const ENTRY_CUSTOM_LABELS = 'me/custom_labels';
