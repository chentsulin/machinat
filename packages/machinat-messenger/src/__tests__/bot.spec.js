import moxy from 'moxy';
import nock from 'nock';
import Machinat from 'machinat';
import Renderer from 'machinat-renderer';
import Queue from 'machinat-queue';
import { Emitter, Engine, Controller } from 'machinat-base';
import WebhookReceiver from 'machinat-webhook-receiver';
import MessengerBot from '../bot';
import MessengerWorker from '../worker';
import { Image, Dialog, QuickReply } from '../component';
import { MESSENGER_NATIVE_TYPE } from '../constant';
import { makeResponse } from './utils';

jest.mock('machinat-base');
jest.mock('machinat-renderer');
jest.mock('machinat-webhook-receiver');

nock.disableNetConnect();

const message = (
  <Dialog quickReplies={<QuickReply title="Hi!" />}>
    Hello
    <b>World!</b>
    <Image url="https://machinat.com/greeting.png" />
  </Dialog>
);

let graphAPI;
const bodySpy = moxy(() => true);

beforeEach(() => {
  Renderer.mock.clear();
  Engine.mock.clear();
  Controller.mock.clear();
  WebhookReceiver.mock.clear();

  graphAPI = nock('https://graph.facebook.com').post('/v3.3/', bodySpy);
  bodySpy.mock.clear();
});

afterEach(() => {
  nock.cleanAll();
});

describe('#constructor(options)', () => {
  it('extends Emitter', () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    expect(bot).toBeInstanceOf(Emitter);
  });

  it('throw if accessToken not given', () => {
    expect(
      () =>
        new MessengerBot({
          pageId: '_PAGE_ID_',
          appSecret: '_SECRET_',
          verifyToken: '_VERIFIY_TOKEN_',
        })
    ).toThrowErrorMatchingInlineSnapshot(
      `"should provide accessToken to send messages"`
    );
  });

  it('throw if pageId not given', () => {
    expect(
      () =>
        new MessengerBot({
          accessToken: '_ACCESS_TOKEN_',
          appSecret: '_SECRET_',
          verifyToken: '_VERIFIY_TOKEN_',
        })
    ).toThrowErrorMatchingInlineSnapshot(
      `"should provide pageId as the identification of resources"`
    );
  });

  it('throw if appSecret not given', () => {
    expect(
      () =>
        new MessengerBot({
          pageId: '_PAGE_ID_',
          accessToken: '_ACCESS_TOKEN_',
          verifyToken: '_VERIFIY_TOKEN_',
        })
    ).toThrowErrorMatchingInlineSnapshot(
      `"should provide appSecret if shouldValidateRequest set to true"`
    );
  });

  it('is ok to have appSecret empty if shouldValidateRequest set to false', () => {
    expect(
      () =>
        new MessengerBot({
          pageId: '_PAGE_ID_',
          accessToken: '_ACCESS_TOKEN_',
          verifyToken: '_VERIFIY_TOKEN_',
          shouldValidateRequest: false,
        })
    ).not.toThrow();
  });

  it('throw if verifyToken not given', () => {
    expect(
      () =>
        new MessengerBot({
          pageId: '_PAGE_ID_',
          accessToken: '_ACCESS_TOKEN_',
          appSecret: '_SECRET_',
        })
    ).toThrowErrorMatchingInlineSnapshot(
      `"should provide verifyToken if shouldVerifyWebhook set to true"`
    );
  });

  it('is ok to have verifyToken empty if shouldVerifyWebhook set to false', () => {
    expect(
      () =>
        new MessengerBot({
          pageId: '_PAGE_ID_',
          accessToken: '_ACCESS_TOKEN_',
          appSecret: '_SECRET_',
          shouldVerifyWebhook: false,
        })
    ).not.toThrow();
  });

  it('set default options', () => {
    expect(
      new MessengerBot({
        pageId: '_PAGE_ID_',
        accessToken: '_ACCESS_TOKEN_',
        appSecret: '_SECRET_',
        verifyToken: '_VERIFIY_TOKEN_',
      }).options
    ).toMatchInlineSnapshot(`
            Object {
              "accessToken": "_ACCESS_TOKEN_",
              "appSecret": "_SECRET_",
              "pageId": "_PAGE_ID_",
              "respondTimeout": 5000,
              "shouldValidateRequest": true,
              "shouldVerifyWebhook": true,
              "verifyToken": "_VERIFIY_TOKEN_",
            }
        `);
  });

  it('covers default options', () => {
    const options = {
      pageId: '_PAGE_ID_',
      appSecret: '_SECRET_',
      accessToken: '_ACCESS_TOKEN_',
      shouldValidateRequest: true,
      shouldVerifyWebhook: true,
      verifyToken: '_VERIFIY_TOKEN_',
      respondTimeout: 9999,
      consumeInterval: 10000,
    };

    expect(new MessengerBot(options).options).toEqual(options);
  });

  it('assemble core modules', () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    expect(bot.engine).toBeInstanceOf(Engine);
    expect(bot.controller).toBeInstanceOf(Controller);
    expect(bot.receiver).toBeInstanceOf(WebhookReceiver);

    expect(Renderer.mock).toHaveBeenCalledTimes(1);
    expect(Renderer.mock).toHaveBeenCalledWith(
      'messenger',
      MESSENGER_NATIVE_TYPE,
      expect.any(Function)
    );

    expect(Engine.mock).toHaveBeenCalledTimes(1);
    expect(Engine.mock).toHaveBeenCalledWith(
      'messenger',
      bot,
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(MessengerWorker),
      []
    );

    expect(Controller.mock).toHaveBeenCalledTimes(1);
    expect(Controller.mock).toHaveBeenCalledWith('messenger', bot, []);

    expect(WebhookReceiver.mock).toHaveBeenCalledTimes(1);
    expect(WebhookReceiver.mock).toHaveBeenCalledWith(
      expect.any(Function),
      expect.any(Function)
    );
  });

  it('pass middlewares from plugins to controller and engine', () => {
    const eventMiddleware1 = () => () => {};
    const eventMiddleware2 = () => () => {};
    const dispatchMiddleware1 = () => () => {};
    const dispatchMiddleware2 = () => () => {};
    const plugins = [
      moxy(() => ({
        dispatchMiddleware: dispatchMiddleware1,
      })),
      moxy(() => ({
        eventMiddleware: eventMiddleware1,
      })),
      moxy(() => ({
        dispatchMiddleware: dispatchMiddleware2,
        eventMiddleware: eventMiddleware2,
      })),
    ];

    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
      plugins,
    });

    expect(Engine.mock).toHaveBeenCalledWith(
      'messenger',
      bot,
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(MessengerWorker),
      [dispatchMiddleware1, dispatchMiddleware2]
    );

    expect(Controller.mock).toHaveBeenCalledWith('messenger', bot, [
      eventMiddleware1,
      eventMiddleware2,
    ]);
  });

  it('issue event & error', async () => {
    const eventIssuerSpy = moxy(() => Promise.resolve());
    Controller.mock.fake(function FakeController() {
      return { eventIssuerThroughMiddlewares: () => eventIssuerSpy };
    });

    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    const eventListener = moxy();
    const errorListener = moxy();
    bot.onEvent(eventListener);
    bot.onError(errorListener);

    expect(bot.receiver.bindIssuer.mock).toHaveBeenCalledTimes(1);
    expect(
      bot.controller.eventIssuerThroughMiddlewares.mock
    ).toHaveBeenCalledTimes(1);
    const finalPublisher =
      bot.controller.eventIssuerThroughMiddlewares.mock.calls[0].args[0];

    const channel = { super: 'slam' };
    const event = { a: 'phonecall' };
    const metadata = { champ: 'Johnnnnn Ceeeena!' };
    const frame = { channel, event, metadata };

    expect(finalPublisher(frame)).toBe(undefined);

    expect(eventListener.mock).toHaveBeenCalledTimes(1);
    expect(eventListener.mock).toHaveBeenCalledWith(frame);

    const [issueEvent, issueError] = bot.receiver.bindIssuer.mock.calls[0].args;

    await expect(issueEvent(channel, event, metadata)).resolves.toBe(undefined);
    expect(eventIssuerSpy.mock).toHaveBeenCalledTimes(1);
    expect(eventIssuerSpy.mock).toHaveBeenCalledWith(channel, event, metadata);

    expect(issueError(new Error('NO'))).toBe(undefined);
    expect(errorListener.mock).toHaveBeenCalledTimes(1);
    expect(errorListener.mock).toHaveBeenCalledWith(new Error('NO'));
  });
});

describe('#send(message, options)', () => {
  let scope;
  beforeEach(() => {
    scope = graphAPI.reply(
      200,
      JSON.stringify([
        makeResponse(200, { message_id: 'xxx', recipient_id: 'xxx' }),
        makeResponse(200, { message_id: 'xxx', recipient_id: 'xxx' }),
        makeResponse(200, { message_id: 'xxx', recipient_id: 'xxx' }),
      ])
    );
  });

  it('resolves null if message is empty', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(bot.send('john', empty)).resolves.toBe(null);
    }
  });

  it('works', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
      plugins: [
        () => ({
          dispatchMiddleware: next => frame => next(frame),
        }),
      ],
    });

    await expect(bot.send('john', message)).resolves.toEqual([
      { code: 200, body: { message_id: 'xxx', recipient_id: 'xxx' } },
      { code: 200, body: { message_id: 'xxx', recipient_id: 'xxx' } },
      { code: 200, body: { message_id: 'xxx', recipient_id: 'xxx' } },
    ]);

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot({ batch: expect.any(String) });
    expect(JSON.parse(body.batch)).toMatchSnapshot();

    expect(scope.isDone()).toBe(true);
  });

  it('works with options', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    await expect(
      bot.send('john', message, {
        messagingType: 'TAG',
        tag: 'TRANSPORTATION_UPDATE',
        notificationType: 'SILENT_PUSH',
        personaId: 'billy17',
      })
    ).resolves.toEqual([
      { code: 200, body: { message_id: 'xxx', recipient_id: 'xxx' } },
      { code: 200, body: { message_id: 'xxx', recipient_id: 'xxx' } },
      { code: 200, body: { message_id: 'xxx', recipient_id: 'xxx' } },
    ]);

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot({ batch: expect.any(String) });
    expect(JSON.parse(body.batch)).toMatchSnapshot();

    expect(scope.isDone()).toBe(true);
  });
});

describe('#createAttachment(message)', () => {
  it('resolves null if message is empty', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      await expect(bot.createAttachment(empty)).resolves.toBe(null); // eslint-disable-line no-await-in-loop
    }
  });

  it('works', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    const scope = graphAPI.reply(
      200,
      JSON.stringify([makeResponse(200, { attachment_id: 401759795 })])
    );

    await expect(
      bot.createAttachment(
        <Image sharable url="https://machinat.com/trollface.png" />
      )
    ).resolves.toEqual({
      code: 200,
      body: { attachment_id: 401759795 },
    });

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot({ batch: expect.any(String) });
    expect(JSON.parse(body.batch)).toMatchInlineSnapshot(`
                  Array [
                    Object {
                      "body": "message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fmachinat.com%2Ftrollface.png%22%7D%7D%7D",
                      "method": "POST",
                      "omit_response_on_success": false,
                      "relative_url": "me/message_attachments",
                    },
                  ]
            `);

    expect(scope.isDone()).toBe(true);
  });
});

describe('#createMessageCreative(message)', () => {
  it('resolves null if message is empty', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      await expect(bot.createMessageCreative(empty)).resolves.toBe(null); // eslint-disable-line no-await-in-loop
    }
  });

  it('works', async () => {
    const bot = new MessengerBot({
      pageId: '_PAGE_ID_',
      accessToken: '_ACCESS_TOKEN_',
      appSecret: '_SECRET_',
      verifyToken: '_VERIFIY_TOKEN_',
    });

    const scope = graphAPI.reply(
      200,
      JSON.stringify([makeResponse(200, { message_creative_id: 938461089 })])
    );

    await expect(bot.createMessageCreative(message)).resolves.toEqual({
      code: 200,
      body: { message_creative_id: 938461089 },
    });

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot({ batch: expect.any(String) });
    expect(JSON.parse(body.batch)).toMatchInlineSnapshot(`
      Array [
        Object {
          "body": "messages=%5B%7B%22text%22%3A%22Hello%22%7D%2C%7B%22text%22%3A%22*World!*%22%7D%2C%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fmachinat.com%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%7D%5D%7D%5D",
          "method": "POST",
          "omit_response_on_success": false,
          "relative_url": "me/message_creatives",
        },
      ]
    `);

    expect(scope.isDone()).toBe(true);
  });
});
