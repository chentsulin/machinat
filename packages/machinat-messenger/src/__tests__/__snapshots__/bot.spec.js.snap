// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render(channel, message, options) works 1`] = `
Object {
  "jobs": Array [
    Object {
      "attachmentAssetTag": undefined,
      "attachmentFileData": undefined,
      "attachmentFileInfo": undefined,
      "channelUid": "messenger:_PAGE_ID_:psid:john",
      "request": Object {
        "body": Object {
          "message": Object {
            "text": "Hello *World!*",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": Object {
            "id": "john",
          },
          "tag": undefined,
        },
        "method": "POST",
        "name": "messenger:_PAGE_ID_:psid:john-1",
        "omit_response_on_success": false,
        "relative_url": "me/messages",
      },
    },
    Object {
      "attachmentAssetTag": undefined,
      "attachmentFileData": undefined,
      "attachmentFileInfo": undefined,
      "channelUid": "messenger:_PAGE_ID_:psid:john",
      "request": Object {
        "body": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://machinat.com/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": undefined,
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": Object {
            "id": "john",
          },
          "tag": undefined,
        },
        "depends_on": "messenger:_PAGE_ID_:psid:john-1",
        "method": "POST",
        "name": "messenger:_PAGE_ID_:psid:john-2",
        "omit_response_on_success": false,
        "relative_url": "me/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "attachmentAssetTag": undefined,
          "attachmentFileData": undefined,
          "attachmentFileInfo": undefined,
          "channelUid": "messenger:_PAGE_ID_:psid:john",
          "request": Object {
            "body": Object {
              "message": Object {
                "text": "Hello *World!*",
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": Object {
                "id": "john",
              },
              "tag": undefined,
            },
            "method": "POST",
            "name": "messenger:_PAGE_ID_:psid:john-1",
            "omit_response_on_success": false,
            "relative_url": "me/messages",
          },
        },
        Object {
          "attachmentAssetTag": undefined,
          "attachmentFileData": undefined,
          "attachmentFileInfo": undefined,
          "channelUid": "messenger:_PAGE_ID_:psid:john",
          "request": Object {
            "body": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://machinat.com/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": Array [
                  Object {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": undefined,
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": Object {
                "id": "john",
              },
              "tag": undefined,
            },
            "depends_on": "messenger:_PAGE_ID_:psid:john-1",
            "method": "POST",
            "name": "messenger:_PAGE_ID_:psid:john-2",
            "omit_response_on_success": false,
            "relative_url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#render(channel, message, options) works 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "8688ea46d23024c65f2e66c2b872fccd774120de9ca4a131cbb14d854ca5a4a6",
  "batch": Any<String>,
}
`;

exports[`#render(channel, message, options) works 3`] = `
Array [
  Object {
    "body": "message=%7B%22text%22%3A%22Hello%20*World!*%22%7D&recipient=%7B%22id%22%3A%22john%22%7D",
    "method": "POST",
    "name": "messenger:_PAGE_ID_:psid:john-1",
    "omit_response_on_success": false,
    "relative_url": "me/messages",
  },
  Object {
    "body": "message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fmachinat.com%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%7D%5D%7D&recipient=%7B%22id%22%3A%22john%22%7D",
    "depends_on": "messenger:_PAGE_ID_:psid:john-1",
    "method": "POST",
    "name": "messenger:_PAGE_ID_:psid:john-2",
    "omit_response_on_success": false,
    "relative_url": "me/messages",
  },
]
`;

exports[`#render(channel, message, options) works with options 1`] = `
Object {
  "jobs": Array [
    Object {
      "attachmentAssetTag": undefined,
      "attachmentFileData": undefined,
      "attachmentFileInfo": undefined,
      "channelUid": "messenger:_PAGE_ID_:psid:john",
      "request": Object {
        "body": Object {
          "message": Object {
            "text": "Hello *World!*",
          },
          "messaging_type": "TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": Object {
            "id": "john",
          },
          "tag": "TRANSPORTATION_UPDATE",
        },
        "method": "POST",
        "name": "messenger:_PAGE_ID_:psid:john-1",
        "omit_response_on_success": false,
        "relative_url": "me/messages",
      },
    },
    Object {
      "attachmentAssetTag": undefined,
      "attachmentFileData": undefined,
      "attachmentFileInfo": undefined,
      "channelUid": "messenger:_PAGE_ID_:psid:john",
      "request": Object {
        "body": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://machinat.com/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": undefined,
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": "TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": Object {
            "id": "john",
          },
          "tag": "TRANSPORTATION_UPDATE",
        },
        "depends_on": "messenger:_PAGE_ID_:psid:john-1",
        "method": "POST",
        "name": "messenger:_PAGE_ID_:psid:john-2",
        "omit_response_on_success": false,
        "relative_url": "me/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "attachmentAssetTag": undefined,
          "attachmentFileData": undefined,
          "attachmentFileInfo": undefined,
          "channelUid": "messenger:_PAGE_ID_:psid:john",
          "request": Object {
            "body": Object {
              "message": Object {
                "text": "Hello *World!*",
              },
              "messaging_type": "TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": Object {
                "id": "john",
              },
              "tag": "TRANSPORTATION_UPDATE",
            },
            "method": "POST",
            "name": "messenger:_PAGE_ID_:psid:john-1",
            "omit_response_on_success": false,
            "relative_url": "me/messages",
          },
        },
        Object {
          "attachmentAssetTag": undefined,
          "attachmentFileData": undefined,
          "attachmentFileInfo": undefined,
          "channelUid": "messenger:_PAGE_ID_:psid:john",
          "request": Object {
            "body": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://machinat.com/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": Array [
                  Object {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": undefined,
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": "TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": Object {
                "id": "john",
              },
              "tag": "TRANSPORTATION_UPDATE",
            },
            "depends_on": "messenger:_PAGE_ID_:psid:john-1",
            "method": "POST",
            "name": "messenger:_PAGE_ID_:psid:john-2",
            "omit_response_on_success": false,
            "relative_url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#render(channel, message, options) works with options 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "8688ea46d23024c65f2e66c2b872fccd774120de9ca4a131cbb14d854ca5a4a6",
  "batch": Any<String>,
}
`;

exports[`#render(channel, message, options) works with options 3`] = `
Array [
  Object {
    "body": "message=%7B%22text%22%3A%22Hello%20*World!*%22%7D&messaging_type=TAG&tag=TRANSPORTATION_UPDATE&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%22john%22%7D",
    "method": "POST",
    "name": "messenger:_PAGE_ID_:psid:john-1",
    "omit_response_on_success": false,
    "relative_url": "me/messages",
  },
  Object {
    "body": "message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fmachinat.com%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%7D%5D%7D&messaging_type=TAG&tag=TRANSPORTATION_UPDATE&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%22john%22%7D",
    "depends_on": "messenger:_PAGE_ID_:psid:john-1",
    "method": "POST",
    "name": "messenger:_PAGE_ID_:psid:john-2",
    "omit_response_on_success": false,
    "relative_url": "me/messages",
  },
]
`;

exports[`#renderAttachment(message) works 1`] = `
Object {
  "jobs": Array [
    Object {
      "attachmentFileData": undefined,
      "attachmentFileInfo": undefined,
      "request": Object {
        "body": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://machinat.com/trollface.png",
              },
              "type": "image",
            },
          },
        },
        "method": "POST",
        "omit_response_on_success": false,
        "relative_url": "me/message_attachments",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "attachment_id": 401759795,
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "attachmentFileData": undefined,
          "attachmentFileInfo": undefined,
          "request": Object {
            "body": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://machinat.com/trollface.png",
                  },
                  "type": "image",
                },
              },
            },
            "method": "POST",
            "omit_response_on_success": false,
            "relative_url": "me/message_attachments",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#renderAttachment(message) works 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "8688ea46d23024c65f2e66c2b872fccd774120de9ca4a131cbb14d854ca5a4a6",
  "batch": Any<String>,
}
`;

exports[`#renderMessageCreative(message) works 1`] = `
Object {
  "jobs": Array [
    Object {
      "request": Object {
        "body": Object {
          "messages": Array [
            Object {
              "text": "Hello *World!*",
            },
            Object {
              "attachment": Object {
                "payload": Object {
                  "attachment_id": undefined,
                  "is_reusable": undefined,
                  "url": "https://machinat.com/greeting.png",
                },
                "type": "image",
              },
              "metadata": undefined,
              "quick_replies": Array [
                Object {
                  "content_type": "text",
                  "image_url": undefined,
                  "payload": undefined,
                  "title": "Hi!",
                },
              ],
            },
          ],
        },
        "method": "POST",
        "omit_response_on_success": false,
        "relative_url": "me/message_creatives",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_creative_id": 938461089,
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "request": Object {
            "body": Object {
              "messages": Array [
                Object {
                  "text": "Hello *World!*",
                },
                Object {
                  "attachment": Object {
                    "payload": Object {
                      "attachment_id": undefined,
                      "is_reusable": undefined,
                      "url": "https://machinat.com/greeting.png",
                    },
                    "type": "image",
                  },
                  "metadata": undefined,
                  "quick_replies": Array [
                    Object {
                      "content_type": "text",
                      "image_url": undefined,
                      "payload": undefined,
                      "title": "Hi!",
                    },
                  ],
                },
              ],
            },
            "method": "POST",
            "omit_response_on_success": false,
            "relative_url": "me/message_creatives",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#renderMessageCreative(message) works 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "8688ea46d23024c65f2e66c2b872fccd774120de9ca4a131cbb14d854ca5a4a6",
  "batch": Any<String>,
}
`;
